model/metrics/dotnet-metrics.yaml

```
@@ -0,0 +1,213 @@
+groups:
+  - id: metric.dotnet.gc.collections
+    type: metric
+    metric_name: dotnet.gc.collections
+    brief: "Number of garbage collections that have occurred since the process has started."
+    note: "This metric reports the same values as calling [`GC.CollectionCount(int generation)`](https://learn.microsoft.com/dotnet/api/system.gc.collectioncount)."
+    instrument: counter
+    unit: "{collection}"
+    stability: experimental
+    attributes:
+      - ref: dotnet.gc.heap.generation
+        requirement_level: required
+
+  - id: metric.dotnet.gc.objects.size
+    type: metric
+    metric_name: dotnet.gc.objects.size
+    brief: "The number of bytes currently allocated on the managed GC heap. Fragmentation and other GC committed memory pools are excluded."
+    note: "This metric reports the same values as calling [`GC.GetTotalMemory(false)`](https://learn.microsoft.com/dotnet/api/system.gc.gettotalmemory)."
+    instrument: updowncounter
+    unit: "By"
+    stability: experimental
+
+  - id: metric.dotnet.gc.memory.total_allocated
+    type: metric
+    metric_name: dotnet.gc.memory.total_allocated
+    brief: >
+      The approximate number of bytes allocated on the managed GC heap since the process has
+      started. The returned value does not include any native allocations.
+    note: "This metric reports the same values as calling [`GC.GetTotalAllocatedBytes()`](https://learn.microsoft.com/dotnet/api/system.gc.gettotalallocatedbytes)."
+    instrument: counter
+    unit: "By"
+    stability: experimental
+
+  - id: metric.dotnet.gc.memory.commited
+    type: metric
+    metric_name: dotnet.gc.memory.commited
+    brief: >
+      The amount of committed virtual memory for the managed GC heap, as
+      observed during the latest garbage collection.
+    note: >
+      This metric reports the same values as calling
+      [`GC.GetGCMemoryInfo().TotalCommittedBytes`](https://learn.microsoft.com/dotnet/api/system.gcmemoryinfo.totalcommittedbytes).
+      Committed virtual memory may be larger than the heap size because it
+      includes both memory for storing existing objects (the heap size) and some
+      extra memory that is ready to handle newly allocated objects in the
+      future.
+    instrument: updowncounter
+    unit: "By"
+    stability: experimental
+
+  - id: metric.dotnet.gc.heap.size
+    type: metric
+    metric_name: dotnet.gc.heap.size
```

```suggestion
    metric_name: dotnet.gc.heap.last_collection_size
```

What do folks think of calling this metric 'last_collection_size' so that in .NET 10 we could add a "dotnet.gc.heap.size" metric which is measuring the instantaneous heap sizes? I expect those live size measurements would have the same generation attribute and report one size for each of gen0,gen1,gen2,loh,poh. I'm not attached to 'last_collection' if there are other name suggestions that convey the measurement is not instantaneous. 'size_at_last_collection'?

If we did this I think we'd want to name other past GC metrics similarly:
dotnet.gc.heap.fragmentation.last_collection_size
dotnet.gc.memory.last_collection_committed_size

---

I think this is a good idea given the potential for a "live" measurement. I slightly lean toward the `size_at_last_collection`. One further option is to try and clear this up in the namespace, so `dotnet.gc.last_collection.heap.size` etc.

---

I also wonder if we want/need to clarify that this is the value **after** collection. We could do that in the metric name, or just in the description/notes. We might want to update the name if we see value in (or may in the future) exposing a metric for the "before" values.

---

If you have a strong preference to have 'after-the-collection' in the metric name and some good naming ideas, it would be nice.

But I believe it's super-common to measure heap after collection and it's expected, so I think that having it in brief is enough.